### ISSUE TITLE
route.iac issue

### ISSUE THREAD

**Issue by davidaranega96** [TUDelft-CNS-ATM/bluesky#261]
Hi guys,
I have been playing with bluesky creating random scenarios and I think there might be an issue with the iac attribute for the Route object. If I am not mistaken this attribute is the index that the aircraft has in the rest of the traffic arrays. 

The problem I see is that when an aircraft is deleted this value is not updated for the other aicrafts and you might end up having "IndexError: index out of bound" erros. In the code below I try to show this by creating 3 ac, deleting one and checking the values for route.iac.
```
import bluesky as bs 
bs.init("sim-detached")
#I create 3 random aircraft
bs.traf.mcre(n=3)
#I add a waypoint to the route for each aircraft.
for ac in range(bs.traf.ntraf):
	bs.traf.ap.route[ac].addwpt(iac = ac, name="RWY123"+str(ac), wptype=4, lat=40.0, lon=0.0)
	print("Aircraft index:	",ac,"	self.iac value inside the route object:	", bs.traf.ap.route[ac].iac)

bs.traf.delete(1)
print("Bluesky step...")
bs.sim.step()

for ac in range(bs.traf.ntraf):
	print("Aircraft index:	",ac,"	self.iac value inside the route object:	", bs.traf.ap.route[ac].iac)
```
Which returns
```
Aircraft index:  0      .iac value inside the route object:      0
Aircraft index:  1      .iac value inside the route object:      1
Aircraft index:  2      .iac value inside the route object:      2
Bluesky step...
Aircraft index:  0      .iac value inside the route object:      0
Aircraft index:  1      .iac value inside the route object:      2
```
Here you can see that the ac index and the iac value are not the same. This will raise an error if for example the plane with index 1 needs to land bluesky/traffic/route.py line 865. (stack.stack("HDG " + str(bs.traf.id[**self.iac**]) + " " + str(wphdg))
https://github.com/TUDelft-CNS-ATM/bluesky/blob/28f2cd4d2ebeb23f0ebfcb434a5ffd8c860077b7/bluesky/traffic/route.py#L69

**Comment by ProfHoekstra** [comment]
Thank you! We’ll have a look at it.

Best regards/Met vriendelijke groeten,
    Jacco Hoekstra


On 10 Feb 2021, at 17:14, David <notifications@github.com> wrote:

﻿

Hi guys,
I have been playing with bluesky creating random scenarios and I think there might be an issue with the iac attribute for the Route object. If I am not mistaken this attribute is the index that the aircraft has in the rest of the traffic arrays.

The problem I see is that when an aircraft is deleted this value is not updated for the other aicrafts and you might end up having "IndexError: index out of bound" erros. In the code below I try to show this by creating 3 ac, deleting one and checking the values for route.iac.

import bluesky as bs
bs.init("sim-detached")
#I create 3 random aircraft
bs.traf.mcre(n=3)
#I add a waypoint to the route for each aircraft.
for ac in range(bs.traf.ntraf):
        bs.traf.ap.route[ac].addwpt(iac = ac, name="RWY123"+str(ac), wptype=4, lat=40.0, lon=0.0)
        print("Aircraft index:  ",ac,"  self.iac value inside the route object: ", bs.traf.ap.route[ac].iac)

bs.traf.delete(1)
print("Bluesky step...")
bs.sim.step()

for ac in range(bs.traf.ntraf):
        print("Aircraft index:  ",ac,"  self.iac value inside the route object: ", bs.traf.ap.route[ac].iac)


Which returns

Aircraft index:  0      .iac value inside the route object:      0
Aircraft index:  1      .iac value inside the route object:      1
Aircraft index:  2      .iac value inside the route object:      2
Bluesky step...
Aircraft index:  0      .iac value inside the route object:      0
Aircraft index:  1      .iac value inside the route object:      2


Here you can see that the ac index and the iac value are not the same. This will raise an error if for example the plane with index 1 needs to land bluesky/traffic/route.py line 865. (stack.stack("HDG " + str(bs.traf.id[self.iac]) + " " + str(wphdg))
https://github.com/TUDelft-CNS-ATM/bluesky/blob/28f2cd4d2ebeb23f0ebfcb434a5ffd8c860077b7/bluesky/traffic/route.py#L69<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_TUDelft-2DCNS-2DATM_bluesky_blob_28f2cd4d2ebeb23f0ebfcb434a5ffd8c860077b7_bluesky_traffic_route.py-23L69&d=DwMCaQ&c=XYzUhXBD2cD-CornpT4QE19xOJBbRy-TBPLK0X9U2o8&r=5JxzVuDw_d0JV-kYyto54d3K2zKNIvniSIeVeSDtif4&m=g-Ey4UbJxYwEOU2r6x1EugP-J1Ce8JyukP3G853XUzU&s=SWBzDsuGlyo393qe2M0-namOyQYGAPOWtY_dK05wC7o&e=>

**Comment by jooste** [comment]
Thanks David! References to aircraft that last longer than a single timestep should indeed be kept by the callsign of the aircraft instead of the index in the traffic arrays. I've fixed this in commit 960bc19.
